### PR TITLE
change: switch to uwsgi

### LINKED ISSUE #684: Implement more robust WSGI server in production
Right now, there are issues with Gunicorn running in production. Namely, when starting up a worker, the worker times out for reasons that are unclear, and Gunicorn goes into a loop where it kills the worker process, only to restart it, time out, and kill it again. After a few tries, a worker _evenutally_ starts.

Why this happens is unclear. Perhaps a switch to uwsgi is in order?

Note: it would be nice to have #681 merged before doing this.

### PR DESCRIPTION
**Pull request by eddieantonio**
Swaps `gunicorn` for `uwsgi` at around version ~2.0 (the "latest" documented version 😭😭😭).

The logs are more helpful now (~~still doesn't tell us when the workers are fully initialized though~~).

```
itwewina_1  | [uWSGI] getting INI configuration from uwsgi.ini
itwewina_1  | *** Starting uWSGI 2.0.19.1 (64bit) on [Thu Mar 11 19:26:45 2021] ***
itwewina_1  | compiled with version: 8.3.0 on 11 March 2021 19:05:16
itwewina_1  | os: Linux-4.19.121-linuxkit #1 SMP Tue Dec 1 17:50:32 UTC 2020
itwewina_1  | nodename: a1997ef5540c
itwewina_1  | machine: x86_64
itwewina_1  | clock source: unix
itwewina_1  | detected number of CPU cores: 2
itwewina_1  | current working directory: /app/CreeDictionary
itwewina_1  | detected binary path: /app/.venv/bin/uwsgi
itwewina_1  | !!! no internal routing support, rebuild with pcre support !!!
itwewina_1  | chdir() to /app/CreeDictionary
itwewina_1  | your memory page size is 4096 bytes
itwewina_1  | detected max file descriptor number: 1048576
itwewina_1  | lock engine: pthread robust mutexes
itwewina_1  | thunder lock: disabled (you can enable it with --thunder-lock)
itwewina_1  | uWSGI http bound on :8000 fd 4
itwewina_1  | uwsgi socket 0 bound to TCP address 127.0.0.1:43465 (port auto-assigned) fd 3
itwewina_1  | Python version: 3.9.1 (default, Jan 12 2021, 16:56:42)  [GCC 8.3.0]
itwewina_1  | Python main interpreter initialized at 0x559a236811b0
itwewina_1  | python threads support enabled
itwewina_1  | your server socket listen backlog is limited to 100 connections
itwewina_1  | your mercy for graceful operations on workers is 60 seconds
itwewina_1  | mapped 416720 bytes (406 KB) for 8 cores
itwewina_1  | *** Operational MODE: preforking+threaded ***
itwewina_1  | WSGI app 0 (mountpoint='') ready in 2 seconds on interpreter 0x559a236811b0 pid: 7 (default app)
itwewina_1  | *** uWSGI is running in multiple interpreter mode ***
itwewina_1  | spawned uWSGI master process (pid: 7)
itwewina_1  | spawned uWSGI worker 1 (pid: 8, cores: 2)
itwewina_1  | spawned uWSGI worker 2 (pid: 10, cores: 2)
itwewina_1  | spawned uWSGI worker 3 (pid: 12, cores: 2)
itwewina_1  | spawned uWSGI worker 4 (pid: 14, cores: 2)
itwewina_1  | spawned uWSGI http 1 (pid: 15)
```

~~I'll continue tinkering with the config before I think this is ready.~~

**EDIT**: it ready

Fixes #684 (hopefully)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -68,3 +68,4 @@ CreeDictionary/CreeDictionary/static/CreeDictionary/js/index.js
 CreeDictionary/CreeDictionary/static/CreeDictionary/css/styles.css
 *.css.map
 *.js.map
+docker-compose.override.yml
```
Comment (eddieantonio) [inline .gitignore:71]: So I can override docker-compose settings locally and _not_ push them to production!

File: CreeDictionary/uwsgi.ini [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+# Automatically buffer POST request bodies.
+# With post-buffering off, the application MUST read EVERY POST body or else
+# the communication protocol could break :/
+post-buffering = true
+
+
+# Hooks
+# See: https://uwsgi-docs.readthedocs.io/en/latest/Hooks.html
+hook-post-app = print:Application loaded!
+hook-accepting-once = exec:echo uWSGI worker accepting -- pid: $PPID
```
Comment (eddieantonio) [inline CreeDictionary/uwsgi.ini:41]: I added these so we can log WHEN the app is loaded (uWSGI loads the app _before_ forking) and when the workers are actually ready to start calling `accept()`!

File: docker/Dockerfile [hunk 3]
```diff
@@ -91,15 +87,21 @@ COPY --from=builder --chown=itwewina /app /app
...
 ENV PATH="${VIRTUAL_ENV}/bin:${PATH}"
 
+# uWSGI will bind HTTP to this port:
+EXPOSE 8000
+# uWSGI stats are accessible on this port
+# I view them with this command:
+#
+#    $ nc 127.0.0.1 9191 | jq
+#
+EXPOSE 9191
```
Comment (eddieantonio) [inline docker/Dockerfile:101]: @andrewdotn I haven't tried uwsgitop with this at all, but this is the port!

File: CreeDictionary/uwsgi.ini (hunk 2 above)
Comment (andrewdotn) [reply CreeDictionary/uwsgi.ini:41]: Should that be `$PPID` or `$PID`?

Comment (eddieantonio) [reply CreeDictionary/uwsgi.ini:41]: `$PPID` -- I added a comment!

File: docker/Dockerfile (hunk 3 above)
Comment (eddieantonio) [reply docker/Dockerfile:101]: See: https://github.com/UAlbertaALTLab/deploy.altlab.dev/pull/13

**Comment by eddieantonio** [comment]
@andrewdotn I incorporated the suggestions and opened up a PR in the deploy documentation: (UAlbertaALTLab/deploy.altlab.dev#13). Let me know if there's anything else that should be done!

**Review by andrewdotn — APPROVED** [review]

**Pull request merged**